design/nfs.md

```diff
@@ -0,0 +1,294 @@
+![Rook Architecture on Kubernetes](../Documentation/media/rook-architecture.png)
+
+## Network File System (NFS)
+
+NFS allows remote hosts to mount file systems over a network and interact with those file systems as though they are mounted locally. This enables system administrators to consolidate resources onto centralized servers on the network.
+
+## Why NFS?
+
+NFS is widely used for persistent storage in kubernetes cluster. Using NFS storage is a convenient and easy way to provision storage for applications.
+An NFS volume allows an existing NFS (Network File System) share to be mounted into the pod.
+The contents of an NFS volume are preserved and the volume is merely unmounted if the pod is stopped/destroyed. This means that an NFS volume can be pre-populated with data, and that data can be “handed off” between pods.
+NFS supports multiple read/write simultaneously so a single share can be attached to multiple pods.
+
+## Design
+With this design Rook is exploring to providing another widely adopted storage option for admins and users of cloud-native environments. This deisign tends to automate NFS starting from its configuration (such as allowed hosts, read/write permissions etc.) to deployment and provisioning. The operations on NFS which cannot be done natively by Kubernetes will be automated.
+NFS doesn’t provide an internal provisioner for kubernetes, so Rook is needed as an external provisioner.
+This design uses NFS-Ganesha server and NFS v4.
+
+### Initial Setup
+
+The flow of creating NFS backed storage in Rook is
+1. The settings are determined and saved in an NFS server CRD (rook-nfs.yaml)
+2. `kubectl create -f rook-nfs.yaml`
+3. When the NFS CRD instance is created, Rook responds to this request by starting the NFS daemon with the required configuration and exports stated in the CRD and creates a service to expose NFS.
+4. NFS volume is ready to be consumed by other pods through a PVC.
+
+### NFS CRD
+
+The NFS CRD spec will specify the following:
+1. NFS server storage backend configuration. E.g., configuration for various storage backends(ceph, ebs, azure disk etc) that will be shared using NFS.
+2. NFS server configuration  
+   The following points are required for configuring NFS server:
+    - export (The volume being exported)
+    - client (The host or network to which the export is being shared)
+    - client options (The options to be used for the client) e.g., read and write permission, root squash etc.
+
+The parameters to configure NFS CRD are:
+
+| Parameter                                 | Description                                                              | Default              |
```

Comment:
Before this table I think it would be helpful to show an example yaml that uses all the properties. Then the reader can visualize the hierarchy

Reply:
will it be fine if this table is after the examples?

Reply:
i was suggesting a pattern of a simple example, followed by the detailed table, followed by more advanced examples. That's fine as well if the table is after all of the examples.

Reply:
something like this?
```yaml
apiVersion: rook.io/v1alpha1
kind: NetworkFileSystem
metadata:
  name: nfs-vol
  namespace: rook
spec:
  replicas: 1
  exports:
  - name: nfs-share
    server:
      accessMode: ReadOnly
      squash: root
      allowedClients:
      - name: host1
        clients: 172.17.0.5
          accessMode: ReadOnly
          squash: root
  persistentVolumeClaim:
    claimName: ebs-claim
```

Reply:
Yes, looks good. We might consider not including the `allowedClients` in the first example though. It could be in a more advanced example later.

Reply:
okay, this sounds good.